### PR TITLE
tree: Centralize ArrayNode logic in one file

### PR DESCRIPTION
**Pull request by CraigMacomber**
## Description

Currently code for node kinds is organized by what part of the implementation logic it is, not by what kind of node it is.

This change reverses that for ArrayNode (with the expectation that if we do this for ArrayNode, the others will follow)

This has a few benefits:

- Saves on dynamic dispatching on NodeKind in some places (ex: copyContent).
- Allows implementation details (like rawFieldNode) to be non-exported.
- Makes it practical to refactor how the implementation of a given node kind works without impacting others (for example if we wanted to make one node kind proxyless).
- Makes it more clear what is needed to add a node kind and easier to fork a node kind (when needed for compat reasons).

This does not fully centralize the array node kind logic since a few things are still scattered. These include at least:

- The node kind enum member. It may be practical to remove this in the future if another down casting/dispatch option is added (like a common base class).
- The schema factory methods to create array nodes (all the real logic has moved, but the entry points still live on schema factory: this seems fine).
- toMapTree's compatibility and hydration logic (there are plans to address this).

## Reviewer Guidance

The review process is outlined on [this wiki page](https://github.com/microsoft/FluidFramework/wiki/PR-Guidelines#guidelines).

Feedback is desired on both this overall pattern, and the actual changes.

As far as code changes go, pretty much everything in here was just a move, with the exception of markContentType being split out from copyContent, and the copyContent call sight in the code moved out of SchemaFactory getting tweaked a bit.

### CODE REVIEW

**Review by Josmithr — APPROVED** [review]
I like this direction. I've been wanting to split up `proxies.ts` in particular for quite a while. Probably worth getting another set of eyes or two on this for consensus, but I'm definitely in favor.

File: packages/dds/tree/src/simple-tree/proxies.ts [hunk 1]
```diff
@@ -1144,3 +724,7 @@ export type FactoryContent =
  * Content which can be inserted into a tree.
  */
 export type InsertableContent = Unhydrated<TreeNode> | FactoryContent;
+
+export function markContentType(typeName: TreeNodeSchemaIdentifier, copy: object): void {
```
[the remark this answers is not in the document]
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/proxies.ts:728]: I don't actually know why we do this. I just keep preserving the behavior.

[the remark this answers is not in the document]
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/proxies.ts:731]: I've added a comment. @noencke Does that seem good?

[the remark this answers is not in the document]
Comment (CraigMacomber) [reply packages/dds/tree/src/simple-tree/proxies.ts:731]: I discussed this with Noah, and I think the new comment is fine. This code is likely to change in the near future anyway.

[the remark this answers is not in the document]
Comment (noencke) [reply packages/dds/tree/src/simple-tree/proxies.ts:731]: Yea it's great!

**Pull request merged**